### ISSUE TITLE
Support type checking with mypy

### ISSUE THREAD

**Issue by rggjan** [matplotlib/matplotlib#20504]
Would be great if matplotlib added type information and support for [type checking](https://www.python.org/dev/peps/pep-0561/) with `mypy`, for example...

**Comment by story645** [comment]
xref: #17991, #13798, #15926

**Comment by jklymak** [comment]
As discussed on the call, Matplotlib may reconsider this after numpy has been through type hinting for a couple of releases.  

In the meantime, if anyone wants to release a separate package that supplies the hinting stubs, matplotlib would be be happy to support, perhaps even on the matplotlib root project (i.e. matplotlib/mpl-hinting) once it was shown to be generally useful.

We will keep this open to track conversation.

**Comment by rggjan** [comment]
Sorry, didn't find the other issues, probably since I was only searching for *mypy*...

**Comment by rggjan** [comment]
Maybe some of this could be taken over:
[Mypy type stubs for NumPy, pandas, and Matplotlib](https://github.com/predictive-analytics-lab/data-science-types)
?

It has the note:
> ⚠️ this project has mostly stopped development ⚠️
> 
> The pandas team and the numpy team are both in the process of integrating type stubs into their codebases, and we don't see the point of competing with them.

**Comment by thomkeh** [comment]
The Pylance team at Microsoft seems to have produced pretty good type stubs: https://github.com/microsoft/python-type-stubs/tree/main/matplotlib

**Comment by tacaswell** [comment]
Also see https://discourse.matplotlib.org/t/i-see-the-type-hints-from-vs-code-but-i-dont-understand-where-they-are-defined-in-the-code-in-github/22786/6?u=tacaswell

**Comment by Dr-Irv** [comment]
> The Pylance team at Microsoft seems to have produced pretty good type stubs for matplotlib: https://github.com/microsoft/python-type-stubs/tree/main/matplotlib

With `pandas`, we created a `pandas-stubs` project https://github.com/pandas-dev/pandas-stubs that evolved as follows:
1. We started with the stubs that Microsoft was maintaining as part of https://github.com/microsoft/python-type-stubs/
2. In that project, we merged in tests from https://github.com/VirtusLab/pandas-stubs to provide a testing framework for the stubs.
3. We met with the leader of  the Microsoft project and VirtusLabs and agreed that the pandas team would start supplying the stubs, that the pandas stubs would be removed from the Microsoft project, and that VirtusLabs would deprecate their stubs.
4. The pylance team now bundles the "official" pandas-stubs with pylance.

I would suggest a similar path for `matplotlib`.  Create a `matplotlib-stubs` project, and work with Microsoft to use their stubs as a starting point.  You might also consider looking at what we are doing in `pandas-stubs` to create a testing methodology for correctness of the stubs.
